### ISSUE TITLE
Primus can't alt+tab in Diablo III (wine). Works with optirun

### ISSUE THREAD

**Issue by XapaJIaMnu** [amonakov/primus#159]
Hey,

When using primus I can't alt tab from Diablo III. The game minimizes fine but refuses to maximize again. The game music is played and the cursor is displayed, but it doesn't actually actually draw anything, it maintains the same picture as the last image before trying to alt tab back (like the browser).

However it works perfectly fine when the game (the battle.net launcher) is started with optirun.

When the game fails to render with primus and I kill it the following is written in the console:

```
primus: fatal: failed to acquire direct rendering context for display thread
```

And sometimes

```
primus: warning: glXSwapBuffers: respawning threads after context change
```

This didn't use to happen until recently (beginning of December) when alt-tabbing an application that was run with primus and wine didn't work at all (the application remained in foreground and you needed to switch to another desktop or hit the "Show desktop" hotkey). I don't know if a change has happened in primus or wine, but I can investigate.  If you have any pointers/suggestions for debugging?

Cheers,

Nick

**Comment by amonakov** [comment]
> alt-tabbing an application that was run with primus and wine didn't work at all

This is the first time such problem is reported to me; with which applications did you observe it?

>  I don't know if a change has happened in primus or wine

In all of 2014, changes to primus happened only on December 28th.

> If you have any pointers/suggestions for debugging?

An easy thing to do first is to `export PRIMUS_VERBOSE=2` before running primus, and see how perf dumps change after alt-tabbing back.

**Comment by XapaJIaMnu** [comment]
Hey,

Then it must be a wine change that exposed the behavior. Could it be related to the compositor I am using? I use kwin (compositing suspended when primus is running).

Here's an output log with PRIMUS_VERBOSE=2

```primus: profiling: upload autodetection: will use PBO path (78 iters)
primus: profiling: readback: 1200x750, 5.3 fps, 94.1% app, 3.4% sleep, 0.1% map, 2.4% wait
primus: profiling: display: 1200x750, 5.3 fps, 96.6% wait, 0.3% upload, 3.1% draw+swap
primus: profiling: readback: 1200x750, 18.0 fps, 87.7% app, 3.7% sleep, 0.4% map, 8.2% wait
primus: profiling: display: 1200x750, 17.8 fps, 87.3% wait, 0.8% upload, 11.9% draw+swap
primus: profiling: upload autodetection: will use PBO path (79 iters)
primus: profiling: readback: 1200x750, 3.1 fps, 96.3% app, 0.6% sleep, 0.1% map, 3.0% wait
primus: profiling: display: 1200x750, 3.3 fps, 96.0% wait, 0.2% upload, 3.8% draw+swap
primus: profiling: display: 1920x1080, 54.0 fps, 24.7% wait, 6.6% upload, 68.7% draw+swap
primus: profiling: readback: 1920x1080, 54.4 fps, 9.2% app, 51.7% sleep, 5.3% map, 33.8% wait
primus: profiling: readback: 1920x1080, 33.7 fps, 48.0% app, 10.8% sleep, 1.1% map, 40.1% wait
primus: profiling: display: 1920x1080, 34.0 fps, 47.8% wait, 4.1% upload, 48.1% draw+swap
primus: profiling: readback: 1200x750, 0.6 fps, 99.7% app, 0.3% sleep, 0.0% map, 0.0% wait
primus: profiling: display: 1200x750, 0.6 fps, 100.0% wait, 0.0% upload, 0.0% draw+swap
primus: profiling: display: 1920x1080, 44.7 fps, 41.7% wait, 5.4% upload, 52.9% draw+swap
primus: profiling: readback: 1920x1080, 45.1 fps, 35.2% app, 37.0% sleep, 4.3% map, 23.6% wait
primus: profiling: readback: 1200x750, 0.6 fps, 99.7% app, 0.2% sleep, 0.0% map, 0.0% wait
primus: profiling: display: 1200x750, 0.6 fps, 100.0% wait, 0.0% upload, 0.0% draw+swap
primus: profiling: display: 1920x1080, 60.0 fps, 0.0% wait, 7.4% upload, 92.6% draw+swap
primus: profiling: readback: 1920x1080, 60.1 fps, 0.0% app, 46.4% sleep, 5.1% map, 48.4% wait
primus: profiling: display: 1200x750, 9.4 fps, 93.2% wait, 0.6% upload, 6.2% draw+swap
primus: profiling: readback: 1200x750, 9.8 fps, 93.0% app, 2.3% sleep, 0.2% map, 4.6% wait
primus: profiling: readback: 1200x750, 2.2 fps, 98.2% app, 0.4% sleep, 0.1% map, 1.4% wait
primus: profiling: display: 1200x750, 2.6 fps, 97.2% wait, 0.1% upload, 2.7% draw+swap
primus: fatal: failed to acquire direct rendering context for display thread
primus: profiling: readback: 1200x750, 0.6 fps, 99.9% app, 0.1% sleep, 0.0% map, 0.0% wait
primus: profiling: display: 1200x750, 0.6 fps, 98.2% wait, 0.0% upload, 1.8% draw+swap
primus: profiling: readback: 1200x750, 0.6 fps, 99.9% app, 0.1% sleep, 0.0% map, 0.0% wait
primus: profiling: display: 1200x750, 0.6 fps, 100.0% wait, 0.0% upload, 0.0% draw+swap
primus: warning: glXSwapBuffers: respawning threads after context change
```
The 1200x750 window is the battle net app one, the full screen is the Diablo III one. I don't really see anything here.

Cheers,

Nick

**Comment by karolherbst** [comment]
I am pretty sure this is related to wine-staging. Somehow some distributions just start shipping it.

**Comment by amonakov** [comment]
Problems in issue #155 look very similar. Can you provide apitrace traces? Wine log dumps with `WINEDEBUG=trace+wgl` might be helpful as well.

**Comment by XapaJIaMnu** [comment]
@amonakov There you go:
[log 1](http://homepages.inf.ed.ac.uk/s1031254/winelog.gz)
[log 2](http://homepages.inf.ed.ac.uk/s1031254/winelog2.gz)

Log one: Started Battle.net launcher, started the game. Alt tabbed. The game crashed, but not primus: fatal message.
Log two: Started the game, alt-tabbed 3 times successfully, the game crashed on the fourth.

I didn't see anything interesting in the logs but I don't really know what to look for.

@karolherbst , I am using arch linux vanilla wine, which as you can see is [vanilla](https://projects.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/wine)

Cheers,

Nick

**Comment by karolherbst** [comment]
I see, I though this issue was related to that, because I started using that. But maybe its just a wine change or something.

**Comment by XapaJIaMnu** [comment]
Just to reiterate, the problem doesn't appear if the application is windowed (or borderless fullscreen windowed) only if it is proper fullscreen.

Cheers,

Nick

**Comment by karolherbst** [comment]
@amonakov  I noticed some fullscreen issues with a lot of other wine applications, too, but never got enough time to create apitraces for that :/ Maybe it also makes sense to bisect wine to see what actually started to cause this.

**Comment by amonakov** [comment]
Please retest after the recent fix for #160

**Comment by XapaJIaMnu** [comment]
Hey,

I just tested it, 5-6 alt tabs, the problem has disappeared. That seems to have fixed it, thank you!

Cheers,

Nick
